### DISCUSSION TITLE
Proper permission to deploy to the Master/Root account

### DISCUSSION THREAD

**Discussion by erictompkins** [gruntwork-io/knowledge-base#490] [Help]
Our AWS account is set up as a multi-account. Master, Prod, Stage, etc. The only thing deployed in Master is `_global/iam-cross-account`. I'm trying to update the modules but I need some recommendations on the access permissions for `master`. It's been quite a while since I first set this up. We are using `aws-vault` for deploying and typically run something like this:

```bash
aws-vault exec prod -- terragrunt apply
```

I'm pretty sure that for security reasons I don't have a profile set up to access the Master account.

I did try to set up `aws-vault` with the AWS root user. While I could run `terragrunt apply` I got a lot of errors.

```text
Error: error reading IAM policy (ARN: arn:aws:iam::aws:policy/AWSSupportAccess): InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: XXXXXXXXX-0527-4774-af73-63792155b99e

  on .terraform/modules/cross_account_iam_roles/modules/iam-policies/main.tf line 85, in data "aws_iam_policy" "AWSSupportAccess":
  85: data "aws_iam_policy" "AWSSupportAccess" {
```

What is the recommended approach to securely deploy to the Master account? 

---

<ins datetime="2022-07-05T16:31:41Z">
  <p><a href="https://support.gruntwork.io/hc/requests/108882">Tracked in ticket #108882</a></p>
</ins>

**Comment by zackproser** [comment]
Hi Eric, 

Could you perhaps be referring to your "Security" account? If so - you'd want to try authenticating via aws-vault to the correct security account profile and then see if you can run `terragrunt plan` there without errors. 

Also, aws-vault expects that you're going to try to authenticate as an IAM user - if you have access to your Security account via your root user - are you able to log in as your root user and check if there are any existing IAM users defined?

**Comment by erictompkins** [comment]
@zackproser Thank you for your reply. Here is our account hierarchy as it was set up by Gruntwork in 2019 when we ordered the Gruntwork Reference Architecture. I believe that Ben Whaley originally set most of it up.

- Master (root account). There is only one user here and it's the root user. Billing is managed in this account. The only current access for other users is a "billing only" IAM role for certain users in the Security account. There doesn't seem to be an IAM user role set up for deploying infrastructure updates.
   - Security - this is where permissions and IAM users are set up. These users are given access to the other accounts with IAM roles.
   - Stage - our stagging environment
   - Prod - our production environment
   - Dev - our development environment.
   - Shared Services - shared services between accounts like Route53 or ECR repos

We can deploy Terraform updates just fine to all of those except for the Master account.  The only Terraform code deployed to Master is the `master/_glabal/iam-cross-account`. It doesn't appear that we have an IAM user set up to deploy to the Master account. 

I'm wondering how Gruntwork usually sets that up. 

If I try to use my Security account I don't have permission to access the S3 bucket where the Terraform state is stored.

There is a `mastercli` IAM user in the master account. It doesn't have a password to log into the console, but it appears to have some permissions for deployment. I can run `terragrunt state list` with that user but if I run `terragrunt plan` I get a lot of errors after the Terraform state is setup up:

```bash
Error: error reading IAM Role (allow-read-only-access-from-other-accounts): InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: 85a20e82-195c-4796-877a-d578661c7643

Error: error reading IAM Role (allow-billing-only-access-from-other-accounts): InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: ffa15129-d349-402c-80ac-0d445c637c37

Error: error reading IAM policy (ARN: arn:aws:iam::aws:policy/AWSSupportAccess): InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: c4098293-22ff-4145-b013-69f0628b08d7

  on .terraform/modules/cross_account_iam_roles/modules/iam-policies/main.tf line 85, in data “aws_iam_policy” “AWSSupportAccess”:
  85: data “aws_iam_policy” “AWSSupportAccess” {

Error: error reading IAM policy (ARN: arn:aws:iam::aws:policy/AWSSupportAccess): InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: 6ba6c054-644f-40f6-a03b-0feb0e4ffdb4

  on .terraform/modules/cross_account_iam_roles/modules/iam-policies/main.tf line 85, in data “aws_iam_policy” “AWSSupportAccess”:
  85: data “aws_iam_policy” “AWSSupportAccess” {

Error: error reading IAM policy (ARN: arn:aws:iam::aws:policy/AWSSupportAccess): InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: 278bbbed-403c-4610-b898-229f4f39b3c6

  on .terraform/modules/cross_account_iam_roles/modules/iam-policies/main.tf line 85, in data “aws_iam_policy” “AWSSupportAccess”:
  85: data “aws_iam_policy” “AWSSupportAccess” {

Error: error reading IAM policy (ARN: arn:aws:iam::aws:policy/AWSSupportAccess): InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: b2a3cf07-f496-42c5-86d9-bb71e6bfb98d

  on .terraform/modules/cross_account_iam_roles/modules/iam-policies/main.tf line 85, in data “aws_iam_policy” “AWSSupportAccess”:
  85: data “aws_iam_policy” “AWSSupportAccess” {

Error: error reading IAM policy (ARN: arn:aws:iam::aws:policy/AWSSupportAccess): InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: c2e1a87c-3c4a-40e4-9a2c-4d9044a2966c

  on .terraform/modules/cross_account_iam_roles/modules/iam-policies/main.tf line 85, in data “aws_iam_policy” “AWSSupportAccess”:
  85: data “aws_iam_policy” “AWSSupportAccess” {

Error: error reading IAM policy (ARN: arn:aws:iam::aws:policy/AWSSupportAccess): InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: f827f8a6-0f98-4035-93d0-45701c9596bb

  on .terraform/modules/cross_account_iam_roles/modules/iam-policies/main.tf line 85, in data “aws_iam_policy” “AWSSupportAccess”:
  85: data “aws_iam_policy” “AWSSupportAccess” {

Error: error reading IAM policy (ARN: arn:aws:iam::aws:policy/AWSSupportAccess): InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: 72940dde-f558-4979-9759-b5620d444427

  on .terraform/modules/cross_account_iam_roles/modules/iam-policies/main.tf line 85, in data “aws_iam_policy” “AWSSupportAccess”:
  85: data “aws_iam_policy” “AWSSupportAccess” {

Error: error reading IAM policy (ARN: arn:aws:iam::aws:policy/AWSSupportAccess): InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: 33112701-9ee1-4432-8d08-49a9c800bc28

  on .terraform/modules/cross_account_iam_roles/modules/iam-policies/main.tf line 85, in data “aws_iam_policy” “AWSSupportAccess”:
  85: data “aws_iam_policy” “AWSSupportAccess” {

Error: error reading IAM policy (ARN: arn:aws:iam::aws:policy/AWSSupportAccess): InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: 89269a5c-ad46-4cc1-8e49-6ef496b58b30

  on .terraform/modules/cross_account_iam_roles/modules/iam-policies/main.tf line 85, in data “aws_iam_policy” “AWSSupportAccess”:
  85: data “aws_iam_policy” “AWSSupportAccess” {

Error: error reading IAM policy (ARN: arn:aws:iam::aws:policy/AWSSupportAccess): InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: 395fe443-44f5-447c-a78e-c202e51f6887

  on .terraform/modules/cross_account_iam_roles/modules/iam-policies/main.tf line 85, in data “aws_iam_policy” “AWSSupportAccess”:
  85: data “aws_iam_policy” “AWSSupportAccess” {

Error: error reading IAM policy (ARN: arn:aws:iam::aws:policy/AWSSupportAccess): InvalidClientTokenId: The security token included in the request is invalid
        status code: 403, request id: 6ca3faf5-d066-416e-a262-3d4e216812b7

  on .terraform/modules/cross_account_iam_roles/modules/iam-policies/main.tf line 85, in data “aws_iam_policy” “AWSSupportAccess”:
  85: data “aws_iam_policy” “AWSSupportAccess” {

ERRO[0012] Hit multiple errors:
Hit multiple errors:
exit status 1
```

**Comment by zackproser** [comment]
Hi Eric, 

Thanks for the detailed update. Here's what I think might be going on: 

> There is only one user here and it's the root user. Billing is managed in this account. 

This suggests to me that this "Master" account was indeed intended to be used only by the root user in unusual circumstances or, to get consolidated billing information. 

I know that for some time we did not offer any modules / landing zone config targeted at this "Root" or "Master" account. I also wonder if your Ref Arch was deployed so long ago that we didn't yet offer: 

https://github.com/gruntwork-io/terraform-aws-service-catalog/tree/master/modules/landingzone/account-baseline-root 

One option would be to try to retrofit that module into your root account. Depending on your current setup this might be painful, so I don't want to paint it as the ideal solution. 

Another idea would be to check if there are other IAM groups available that have more permissive policies attached to them, that you could add your `mastercli` user to. 

All that said, please let me know if this remains a blocker for you - we may be able to do something for you by way of an updated Ref Arch experience if that would be helpful!

**Comment by erictompkins** [comment]
Thank you for the information about the baseline root module. I will look into it further.

Regarding other IAM groups for the master account, it sounds like I may have to add them through the admin UI instead of with Terraform since I can't deploy with Terraform at all, correct?

If that's the case, once I manually add the groups to the `mastercli` user, what is the recommended way to then associate that with the Terraform code? Or is that even necessary?

My concern about an updated Ref Arc option is that when we got the original one we still added a lot on top of it for our needs. I'm concerned about if our current infrastructure could be utilized still or if we'd have to start fresh, which would not be ideal.

**Comment by zackproser** [comment]
Hi Eric, 

Double checked with my colleagues and they had this to say: 

We really don't do anything with the root account vis a vis Terraform config / IaC and we generally recommend that you do not hook it up to the same IaC config you're using for your security account, for instance. The reason being is that the root user is just too powerful - there are scenarios where someone with your root user credentials could deploy resources that wouldn't be tracked by the org but would be billed to the org, which is not desirable. 

From spot-checking your initial query, our guess is that this may be some special customization that you set up on your own (possibly under Ben's direction?) but it would be hard for us to trace down exactly what changes were made and, what, if any, changes should be made to bring this back into line with one of our standard Ref Arch deployments. 

You actually need to run  `aws-vault exec` with the `--no-session` when using the root user credentials, as certain actions are restricted when using a federated session which is what AWS Vault defaults to.

I can understand why you wouldn't want to start anew if you performed lots of customizations on your delivered Ref Arch, but keep my offer in mind in case a latest and greatest Ref Arch (we'd definitely give you a huge break on it) would in any way help you folks out.
